### PR TITLE
Point "deref coercions" links to new book

### PR DESCRIPTION
**Pull request by ruuda**
Currently the link on doc.rust-lang.org is semi-broken; it links to a page that links to the exact page in the first edition in the book, or to the index of the second edition of the book. If the second editions
is the recommended one now, we should point the links at that one. (In the mean time, the links have been updated to point directly to the first edition of the book, but that hasn't made it onto
the stable channel yet.) By the time this commit makes it onto the stable channel, the second edition of the book should be complete enough. At least the part about deref coercions is.

r? @steveklabnik

### CODE REVIEW

**Comment by ruuda** [comment]
It does generate tidy errors:

```
tidy error: rust/src/libcore/ops/deref.rs:16: line longer than 100 chars
tidy error: rust/src/libcore/ops/deref.rs:74: line longer than 100 chars
tidy error: rust/src/libstd/lib.rs:206: line longer than 100 chars
```
Is there anything that can be done about that?

**Comment by aidanhs** [comment]
@ruuda thanks for the PR! We'll check in to make sure @steveklabnik or another reviewer gets to it soon.

There are allowances made for URLs (`src/tools/tidy/src/style.rs`, `long_line_is_ok`), but I guess it doesn't pick up relative paths so I suppose that would need to be fixed as well. I don't know if there's already an issue open, but I think it'd just mean altering `w.starts_with("http://") || w.starts_with("https://")` to also allow for `../`.

**Comment by ruuda** [comment]
@adrianhs Thank you for pointing me in the right direction. I’ve opened #43632. With that applied, this PR now passes the tidy check.

File: src/libstd/lib.rs [hunk 1]
```diff
@@ -80,8 +80,8 @@
 //!
 //! Note the documentation for the primitives [`str`] and [`[T]`][slice] (also
 //! called 'slice'). Many method calls on [`String`] and [`Vec<T>`] are actually
-//! calls to methods on [`str`] and [`[T]`][slice] respectively, via [deref
-//! coercions].
+//! calls to methods on [`str`] and [`[T]`][slice] respectively, via
+//! [deref-coercions].
```
Comment (sanxiyn) [inline src/libstd/lib.rs]: Why this change? (Hyphen is added to the link text.)

Comment (ruuda) [inline src/libstd/lib.rs]: I added it because the tidy check does not deal with spaces in named links, not realising that it was the short form. Fixed.

**Comment by arielb1** [comment]
Tidy errors: (you can rebase your PR on top of master to ignore them?)
```
[00:03:12] tidy error: /checkout/src/libcore/ops/deref.rs:16: line longer than 100 chars
[00:03:12] tidy error: /checkout/src/libcore/ops/deref.rs:74: line longer than 100 chars
[00:03:12] tidy error: /checkout/src/libstd/lib.rs:206: line longer than 100 chars
[00:03:13] some tidy checks failed
[00:03:13] 
[00:03:13] 
```

**Comment by ruuda** [comment]
@arielb1: Rebased, this now passes the tidy check.

**Comment by arielb1** [comment]
Hi @steveklabnik - I think this is ready for review.

**Comment by QuietMisdreavus** [comment]
@bord r+ rollup

Thanks!

**Comment by QuietMisdreavus** [comment]
@bors r+ rollup

Thanks! (oops, i misspelled the bot's name in my last one `>_>`)
